### PR TITLE
interfaces: fix udev tagging for hooks

### PR DESCRIPTION
**Pull request by zyga**
This branch aims to fix udev tagging to support hooks in addition to apps that
were added earlier.

There are three main patches here:

 - add `spec.TagDevice` method to `udev.Specification`, this automatically tags
   both apps and hooks
 - switch non-common interfaces to `TagDevice`, with exception of some special
   uses of `AddSnippet`
 - switch common interfaces to `TagDevice`, with some trivial changes to how
   that is defined.

The only remaining uses of `spec.AddSnippet` are special-cases that add large
chunk of arbitrary udev code that is not specifically aiming to tag a device.

This passes local unit testing, no testing was done on any hardware.

Base on initial feedback I also included changes to format udev rules
differently.  They are now preceded by the snapd interface name (as a comment).
The order of entries is also different, now being sorted by TAG += if one
exists (this allows us to see everything affecting particular command easily)

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: interfaces/builtin/common.go [hunk 1]
```diff
@@ -147,13 +146,8 @@ func (iface *commonInterface) SecCompConnectedPlug(spec *seccomp.Specification,
...
 func (iface *commonInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.Plug, plugAttrs map[string]interface{}, slot *interfaces.Slot, slotAttrs map[string]interface{}) error {
-	old := "###CONNECTED_SECURITY_TAGS###"
-	if iface.connectedPlugUDev != "" {
-		for appName := range plug.Apps {
-			tag := udevSnapSecurityName(plug.Snap.Name(), appName)
-			snippet := strings.Replace(iface.connectedPlugUDev, old, tag, -1)
-			spec.AddSnippet(snippet)
-		}
+	for _, rule := range iface.connectedPlugUDev {
+		spec.TagDevice(rule)
```
Comment (stolowski) [inline interfaces/builtin/common.go]: Nice.

File: interfaces/builtin/hidraw.go [hunk 2]
```diff
@@ -172,14 +172,11 @@ func (iface *hidrawInterface) UDevConnectedPlug(spec *udev.Specification, plug *
...
-			spec.AddSnippet(fmt.Sprintf("SUBSYSTEM==\"hidraw\", KERNEL==\"%s\", TAG+=\"%s\"", strings.TrimPrefix(path, "/dev/"), tag))
-
-		} else {
-			spec.AddSnippet(udevUsbDeviceSnippet("hidraw", usbVendor, usbProduct, "TAG", tag))
-		}
+	if hasOnlyPath {
+		spec.TagDevice(fmt.Sprintf(`SUBSYSTEM=="hidraw", KERNEL=="%s"`, strings.TrimPrefix(path, "/dev/")))
+	} else {
+		spec.TagDevice(fmt.Sprintf(`IMPORT{builtin}="usb_id"
+SUBSYSTEM=="hidraw", SUBSYSTEMS=="usb", ATTRS{idVendor}=="%04x", ATTRS{idProduct}=="%04x"`, usbVendor, usbProduct))
```
Comment (stolowski) [inline interfaces/builtin/hidraw.go]: I'm not totally sure it's more readable than calling udevUsbDeviceSnippet helper here to be honest. Anyway, not biggie but if you decide not to use udevUsbDeviceSnippet, then perhaps udevUsbDeviceSnippet can be removed from utils?

File: interfaces/udev/spec_test.go [hunk 3]
```diff
@@ -80,6 +86,24 @@ func (s *specSuite) TestAddSnippte(c *C) {
...
 }
 
+func (s *specSuite) TestTagDevice(c *C) {
+	// TagDevice acts in the scope of the plug/slot (as appropriate) and
+	// affects all of the apps and hooks related to the given plug or slot
+	// (with the exception that slots cannot have hooks).
+	iface := &ifacetest.TestInterface{
+		InterfaceName: "test",
+		UDevConnectedPlugCallback: func(spec *udev.Specification, plug *interfaces.Plug, plugAttrs map[string]interface{}, slot *interfaces.Slot, slotAttrs map[string]interface{}) error {
+			spec.TagDevice(`kernel="voodoo"`)
```
Comment (stolowski) [inline interfaces/udev/spec_test.go]: Could you please extend this test to have call spec.TagDevice twice?

File: interfaces/builtin/hidraw.go (hunk 2 above)
Comment (jdstrand) [reply interfaces/builtin/hidraw.go]: I agree, either use udevUsbDeviceSnippet where you can or remove it from utils.go.

File: interfaces/builtin/ofono.go [hunk 4]
```diff
@@ -340,14 +326,18 @@ func (iface *ofonoInterface) DBusPermanentSlot(spec *dbus.Specification, slot *i
...
+	/*
+	   1.Linux modem drivers set up the modem device /dev/modem as a symbolic link
+	     to the actual device to /dev/ttyS*
+	   2./dev/socket/rild is just a socket, not device node created by rild daemon.
+	     Similar case for chnlat*.
+	   So we intetionally skipped modem, rild and chnlat.
+	*/
+	spec.TagDevice(`KERNEL=="tty[A-Z]*[0-9]*|cdc-wdm[0-9]*"`)
+	spec.TagDevice(`KERNEL=="tun"`)
+	spec.TagDevice(`KERNEL=="tun[0-9]*"`)
```
Comment (jdstrand) [inline interfaces/builtin/ofono.go]: Oh, this `tun[0-9]*` tag is actually an error. I realize this PR didn't introduce this, but we should fix it. See https://github.com/snapcore/snapd/pull/4031

File: interfaces/builtin/hidraw.go (hunk 2 above)
Comment (zyga) [reply interfaces/builtin/hidraw.go]: I dropped it entirely now. I think the result is much more readable.

File: interfaces/builtin/ofono.go (hunk 4 above)
Comment (zyga) [reply interfaces/builtin/ofono.go]: Dropped

**Comment by zyga** [comment]
The udev rules cannot contain trailing comments:

`foo # comment`

This is not allowed. I'll update the PR after lunch.

**Comment by zyga** [comment]
Just for the record, this does contain the formatting changes that were requested. Jamie just didn't see it when he made that comment above. We discussed this on IRC yesterday.

**Comment by zyga** [comment]
@pedronis yes (now)

**Pull request merged**